text/0000-improved-ember-registry-apis.md

```diff
@@ -0,0 +1,556 @@
+
+We also introduce `FactoryTypeIdentifier` to distinguish between injections for *all factories of a given type* (`FactoryTypeIdentifier`) and *specific factories* (`Identifier`):
+
+```ts
+interface FactoryTypeIdentifier {
+  type: string;
+}
+```
+
+As is the case in the microsyntax-based design, these factory type injections may not be namespaced.
+
+### `Resolver`
+
+Ember’s `Resolver` function is public API, designed to be customized and overridden. To support backwards compatibility with existing custom resolvers, we introduce a `schemaVersion` key to the `Resolver` API, which may be `undefined` or an integer value. If `schemaVersion` is `undefined` or `0`, the legacy behavior of the resolver is maintained:
+
+```ts
+interface ResolverV0 {
+  schemaVersion?: 0;
+  resolve(fullName: string);
+}
+```
+
+If `schemaVersion` is `1`, the `resolve` function has a new signature, using the `Identifier` type introduced above:
+
+```ts
+interface ResolverV1 {
+  schemaVersion: 1;
+  resolve(identifier: Identifier)
+}
+```
+
+The public API is *either* of the Resolver variants:
+
+```ts
+type Resolver = ResolverV0 | ResolverV1;
+```
+
+The other registry API changes may also be implemented in terms of the new resolver `schemaVersion` check.
+
+This design is completely backwards compatible: it continues working exactly as it does today, with users opting *into* the change by upgrading their version of `ember-resolver`, which will implement the new design proposed by this RFC as a breaking change.
+
+There is also some prior art for this kind of schema versioning: [a similar “stamp”](https://github.com/emberjs/ember.js/pull/9994) was used for accomodating different behaviors when using the modules-based resolver vs. not (back in 2014)!
+
+### Owner APIs
+
+The owner APIs all change to use `Identifier` or `FactoryTypeIdentifier` instead of strings. 
+
+Throughout, for the purposes of registration lookup, we use deep object value equality, *not* object identity. This specifically applies to:
+
+- `Owner.factoryFor`
+- `Owner.lookup`
+- `Owner.registerOptions`
+- `Owner.registerOptionsForType`
+- `Owner.registeredOption`
+- `Owner.registeredOptions`
+- `Owner.registeredOptionsForType`
+- `Owner.resolveRegistration`
+
+####  Options
+
+Some `Owner` registry APIs take lookup or registration options. This RFC does not propose changing these options in any way, but their definitions are provided here for completeness:
+
+```ts
+export interface LookupOptions {
+  singleton?: boolean;
+  instantiate?: boolean;
+  source?: string;
+  namespace?: string;
+}
+
+interface RegisterOptions {
+  singleton?: boolean;
+  instantiate?: boolean;
+}
+```
+
+#### `Factory` and `FactoryManager`
+
+Some APIs refer to *factories* and *factory managers*. The API for these classes is the same as it was previously, with the exception that the `fullName` and `normalizedNames` are deprecated:
```

Comment:
We should replace these APIs with _something_ (it is hard to deprecate them otherwise). Maybe:

```ts
interface Factory<T, C extends FactoryClass | object = FactoryClass> {
  class?: C;
  identifier: FactoryIdentifier,
  fullName?: string; // DEPRECATED
  normalizedName?: string; // DEPRECATED
  create(props?: { [prop: string]: any }): T;
}
```

Reply:
Ah, thank you—I hadn't thought of that. 😅 

→ 64310896